components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Encoders/JsonEncoderEx.cs

```diff
@@ -1320,6 +1320,56 @@ public enum JsonEncoding {
             }
         }
 
+        /// <summary>
+        /// Write multi dimensional array in structure.
+        /// </summary>
+        private void WriteStructureMatrix(
+            string fieldName,
+            Matrix matrix,
+            int dim,
+            ref int index,
+            TypeInfo typeInfo) {
+            // check the nesting level for avoiding a stack overflow.
+            if (_nestingLevel > Context.MaxEncodingNestingLevels) {
+                throw ServiceResultException.Create(
+                    StatusCodes.BadEncodingLimitsExceeded,
+                    "Maximum nesting level of {0} was exceeded",
+                    Context.MaxEncodingNestingLevels);
+            }
+
+            _nestingLevel++;
+
+            try {
+                var arrayLen = matrix.Dimensions[dim];
+                if (dim == matrix.Dimensions.Length - 1) {
+                    // Create a slice of values for the top dimension
+                    var copy = Array.CreateInstance(
+                        matrix.Elements.GetType().GetElementType(), arrayLen);
+                    Array.Copy(matrix.Elements, index, copy, 0, arrayLen);
+                    // Write slice as value rank
+
+#if false // TODO is this really needed when we use the JsonWriter?
```

Comment:
Remove it. But we should have tests... any tests from the reference implementation we can add with this PR?

Reply:
I'll take a look

Reply:
These is no tests as far as I can see in the reference implementation